[1.4] TileIO rewrite

Solxanich opened the pull request:
### What is the new feature?
Upfront, this ended up being a fairly large amount of changes.

Resolve feature request made in issue #37 :
1) Add placeholder walls to preserve modded walls through accidental world loading without mod (implementation drafted by and credited to Direwolf420)
2) Add placeholder chests to preserve chest contents through accidental world loading without mod

Further, this also required a redesign in order to minimize impact on file size and inconsistent saving/loading methods for placeholders. This redesign encompasses the following changes:
1) Switching tile placeholders location-specific data from being indexed by modifying tile frames to instead being indexed by a combination of its position (simplified to a single integer) and an integer to integer dictionary. This facilitated for chests and walls to be implemented consistent with tiles, and is readily expandable to other modded instances such as liquids and wires through a dedicated class (PosIndexing). 
2) With the above switch made, per Direwolf420's recommendation, then refactored the unloadedTile class to be unloadedInfo with multiple subclasses for walls, tiles, and chests, so that unique data could be stored with respect to each without affecting each other. Further, with the switch we no longer need to record tile frame data, leading to some minimal file save size reduction.
3) With the above two redesigns complete, one can then reduce the presence of the previous placeholder tile implementation, as it becomes unnecessarily redundant as direct changes in tile types can be readily made without re-indexing.
4) To resolve file size increase concerns raised by Direwolf420, leverage that the dictionary is 
created as ordered integers for the keys and the existing re-occurence code in TileIO so that future calls to the dictionary by position can round down to get the same effective entry instead of creating a new entry for each. Also required implementing a custom dictionary clearing after all tiles that can be restored are restored. Overall, the file size when working with ExampleMod appears to be within control after adding all changes.
5) Adding additional flags save space to the TileIO ReadModTile/WriteModTile to facilitate for location-based chest style preservation, and possible future flags for modded liquid amount or mod-specific custom data. This also includes adding a legacy compat flag to allow for loading previous worlds into the new system without breaking the save. 

### Why should this be part of tModLoader?
This PR resolves the longstanding (5 yrs) issue of placeholders for walls, and preserving chest contents. Mods like Calamity (abyss), Mod of Redemption (underground lab), and Ancients Awakened heavily add modded chests and modded walls, and consequently these objects become quite important to the mod's integrity of play. 
By implementing these placeholders, players who load the world without the requisite mods will not be inadvertently punished by losing access to modded chests contents, or functionality dependent upon the mod's walls.

### Are there alternative designs?
Not aware of alternative designs that would meet needs, however am open to discussion.

On a similar note, this is a not a complete comprehensive PR. There are additional improvements/deficiencies that could be dealt with:
1) The legacy compat flag depends on at least one modded tile existing in the previous version of the save. This would mean that any world deficient of modded tiles but possessing modded walls would break on loading into new system, however *my opinion* is that this is a edge case to ignore.
2) Further re-factor boiler plate code in UnloadedTilesWorld, in particular in the load function.
3) Possible refactor containers and tile entities to take advantage of feature, where applicable
3) Add modded liquids, modded wires, mod-specific custom data etc. as required based on modded walls implementation

### Sample usage for the new feature
n/A at this time. Feature is intended to be independent to mod implementation.

### ExampleMod updates
Updated example PlayerInventory modifications in ExampleMod to include additional items to facilitate testing, and improve demonstration of adding different types of objects. Not required as part of feature implementation.

direwolf420 commented:
Didn't expect a PR so soon. Concider setting it as a draft instead as in its current state its not really safe to merge.

As I pointed out in the linked issue, I think an important milestone towards this PR is figuring out an efficient and unified way of handling per-tile data so that it is also usable in other mods easily without them having to write hundreds of lines of boilerplate

Solxanich commented:
Thanks Direwolf - to confirm, the particular save-breaking cases of concern (as far as safety goes) are:
1) Loading a save with pre-existing unloadedTiles, as this won't work with the changes to how tile info is saved 
2) TileIO has the following lines that would lead to breaking changes if unloaded placeholders are created on a save during first-time loading of this change (ie 1.3.5 -> 1.4 with mods missing would break) due to new save data: Line 88 & 89
`["IsChest"] = TileID.Sets.BasicChest[type], // Flag Chests for preserving Chest Tile data 
["IsSolid"] = Main.tileSolid[type], // Flag non-solid blocks for liquids, player movement, etc.` 

And the change dataflags was removed from this PR, so that isn't a concern at this time

To confirm your point on an additional improvement on handling per-tile data such that mod writers can leverage it, would that not be more appropriate to address as a separate PR that improves on the current system? As is, this PR already notes it is not a comprehensive PR as there are outstanding items that can be improved on that have additional impacts and should be separated out (changing how mods are written or breaking all saves by modifying flags being examples)

Solxanich commented:
Nvm the latter, apologies for my abrasion; I see a few possible opportunities to explore further:
1) I'll look into making the PosIndexer more general so that mods could call it to store data. Were you thinking more of attaching data to the tileInfo or custom objects like Calamity's recharge station? I also guess I'll take a look at the containers and tile entities that have otherwise been untouched through all this if I'm tackling a more generic method.
2) I figured out the backwards compat - the added field IsSolid in the tile overhead tagcompound would not exist on previous saves, so thus a quick ContainsKey is sufficient to if statement around any flags based issues. I'll look to implement.

direwolf420 commented:
1: No, that's tile entities. I meant a way to attach any data to a set of coordinates (most often tile related), e.g. a mod would want to flag a tile as "protected" for mods that add multiplayer protection on tiles (Minecraft has mods like this, albeit for regions), or flags in general (There is a mod called worldtilesflags but it doesnt save tiles efficiently, doesn't support cross mod compatibility, and makes worlds load/save 3-10 seconds longer). And yes, if that were to be a code feature of tml, it would probably be a separate PR is my bet. But I don't know if there are any API devs wanting to tackle this.

2: Nice

Solxanich commented:
Hmm. Just some discussion on my thoughts on what I would do with the custom data implementation for what you are describing:
1) I would break up and/or refactor UnloadedPosIndexing to better handle the info types (it would be best to generalize it for any datatype of 'info') and make it possible for mod's to call SaveInfoToPos, each referencing their own subclass of UnloadedInfo, and own appropriate Infos list, and matching mapping Dictionary. I would assume writers would populate the 'info' themselves
2) I would refactor the info/canRestore groups in UnloadedTileWorld to make it generic for any info type, similar to what would be done with UnloadedPosIndexing so that it could be called to handle their Infos group on reload
3) Again, similar refactor on data cleanup code in the Load function.
4) Leave any restoration/usage of the infos group to the mod writer to decide. 

This would expose all the elements needed to store by a PosID to a mod author imo; if you are content with that description, I can create an issue requesting exposing Position Indexing to authors, with the above summary of how *I* might implement it, and let that independent feature prioritize itself.

Solxanich commented:
I'm thinking this should be reviewed soon;
Given it's been a bit, I've also added some stuff from what was going to be part 2 of the rework: implementing structural lines for fallback types and placeholder purges as part #302; refactoring the load function in UnloadedTilesWorld.

Assuming this doesn't become conflicting with modsystems or other major changes, this should be mergeable.

Solxanich commented:
Kittenchilly + Direwolf comments:
1) Expand to include semi-solid, planters, dressers, and other containers

Will implement dressers and semi-solid, plus structure code for others.

Mirsario commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedChestInfo.cs:
```diff
@@ -0,0 +1,29 @@
+{
+	internal class UnloadedChestInfo : UnloadedInfo 
+	{
+		internal byte ChestStyle; 
```
This and many other fields should be in camelCase.

Rartrin replied:
Either do like Mirsario said and also add `readonly` to it.
`internal byte readonly chestStyle;`

Or just make it a readonly auto property and leave it UpperCamelCase (considering that it's non-private).
`internal byte ChestStyle { get; }`

Solxanich replied:
Fixed this line to be internal readonly byte chestStyle.

Mirsario commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedDresser.cs:
```diff
@@ -0,0 +1,59 @@
+	[Autoload(false)] // need two named versions
+	public class UnloadedDresser : ModTile
+	{
+		public override string Name{get;}
```
```suggestion
		public override string Name { get; }
```

Solxanich replied:
Fixed all instances of spacing error.

Mirsario commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedTilesSystem.cs:
```diff
@@ -7,111 +7,180 @@ namespace Terraria.ModLoader.Default
 
-			if (canRestoreFlag) {
-				RestoreTiles(canRestore);
+			// Cleanup data remnants
+			tileUpdater.cleanupMaps();
+			tileUpdater.cleanupInfos();
+			wallUpdater.cleanupMaps();
+			wallUpdater.cleanupInfos();
+			chestUpdater.cleanupMaps();
+			chestUpdater.cleanupInfos();
```
All methods must be PascalCase, read the style guide - https://github.com/tModLoader/tModLoader/wiki/tModLoader-Style-Guide

Solxanich replied:
Noted - it appears I switched along the way for that class.
Fixed the casing.

Mirsario commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedWallInfo.cs:
```diff
@@ -0,0 +1,23 @@
+namespace Terraria.ModLoader.Default
+{
+	internal class UnloadedWallInfo : UnloadedInfo 
+	{
+		
+		public UnloadedWallInfo(string modName, string name, ushort fallbackType = 0) :base(modName, name,fallbackType) {
```
```suggestion
	{
		public UnloadedWallInfo(string modName, string name, ushort fallbackType = 0) :base(modName, name,fallbackType) {
```

Rartrin replied:
Add a space between `name,` and `fallbackType`.

Solxanich replied:
Resolved spacing issues in all info subclasses.

Mirsario commented on patches/tModLoader/Terraria/Terraria.csproj.patch:
```diff
@@ -82,6 +82,12 @@
      <EmbeddedResource Include="Localization/Content/**" />
      <EmbeddedResource Include="Microsoft/**" />
    </ItemGroup>
++  <ItemGroup>
++    <None Remove="ModLoader\Default\UnloadedDresser.png" />
++  </ItemGroup>
+   <Target Name="EditBin" AfterTargets="Build">
```
Uh.

Solxanich replied:
Fixed. Removed lines; not sure how that got in there, but they're gone

Solxanich replied:
Re: fixed. File no longer a part of PR

Mirsario reviewed (changes requested):
Style review mostly.

Mirsario commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedPosIndexing.cs:
```diff
@@ -0,0 +1,60 @@
+					pendingID++;
+				if (pendingID == infos.Count)
+					infos.Add(info);
+				else
+					infos[pendingID] = info;
+			}
+			modSystem.chestInfoMap[PosID] = pendingID;
+		}
+
+		public void SaveTileInfoToPos(UnloadedTileInfo info) {
+			UnloadedTilesSystem modSystem = ModContent.GetInstance<UnloadedTilesSystem>();
+			var infos = modSystem.tileInfos;
+			int pendingID = infos.IndexOf(info);
+			if (pendingID < 0) {
+				pendingID = 0;
+				while (pendingID < infos.Count && infos[pendingID] != null)
+					pendingID++;
+				if (pendingID == infos.Count)
+					infos.Add(info);
+				else
+					infos[pendingID] = info;
+			}
+			modSystem.tileInfoMap[PosID] = pendingID;
+		}
+
+		public void SaveWallInfoToPos(UnloadedWallInfo info) {
+			UnloadedTilesSystem modSystem = ModContent.GetInstance<UnloadedTilesSystem>();
+			var infos = modSystem.wallInfos;
+			int pendingID = infos.IndexOf(info);
+			if (pendingID < 0) {
+				pendingID = 0;
+				while (pendingID < infos.Count && infos[pendingID] != null)
+					pendingID++;
+				if (pendingID == infos.Count)
+					infos.Add(info);
+				else
+					infos[pendingID] = info;
+			}
+			modSystem.wallInfoMap[PosID] = pendingID;
+		}
```
These 3 methods can be turned into 1 with the right parameters.

Rartrin replied:
Alternatively, a private method to handle all the pendingID stuff. Though, if things could be generalized to use generics, that would be good too. Probably would need some changes to UnloadedTilesSystem.

Solxanich replied:
So, I refactored all the info subclasses, as since the original refactoring of that, a lot has changed.
I was able to reduce it down to one info class again which meant I could mostly clean this to one method. This should be 'better', although I can't say I feel is now perfect.

Solxanich replied:
Alright, got a burst of inspiration and am happy now with the end version of this file.
Very simple class with one method, and short method too.

Solxanich replied:
@Mirsario - are you content with the changes made to this file to clean it up?
If so, I'd like to resolve the change and move towards merging

Rartrin commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedInfo.cs:
```diff
@@ -0,0 +1,26 @@
+{
+	internal class UnloadedInfo
+	{
+		public string modName;
+		public string name;
+		internal ushort FallbackType;
```
Make these readonly properties and use UpperCamelCase.

Solxanich replied:
Not sure about this one. 
Readonly wouldn't work with current subclasses for the different tile types constructing base values afaik (namely looking at chests while thinking this). I've swapped from public to internal, but that's not getting to what you're looking for I presume.

Per Mirsario's comment, these are attributes so camelCase not Upper.

Rartrin replied:
I was mainly talking about the public ones fields, the conventions for attributes is to use public properties rather than public fields.
As for `internal`, everywhere I've seen them used, they follow nearly the same conventions as `public` because they are practically the same except you have to whitelist which external assemblies can use them. In this case, these attributes are meant to be used outside of the class itself so a property would just make more sense then a field and be cleaner.

As for readonly, I understand what you mean about not being able to be readonly. I wasn't able to see much else that so I was just putting it out there as something to check.

Does make me wonder if there would be a better way to handle it from an object oriented approach to make it more protected. Setting a base field in an extending class outside of its constructor is generally a suboptimal approach.

Solxanich replied:
Thanks for the education there Rartrin; I didn't quite understand the difference between them all as I've only started to do OOP this year. Makes a lot more sense now; and I'm all for safe coding practices.

I made them readonly when I refactored the UnloadedInfo back down to one class, so that's been fixed. Realized I could just store a customData tagcompound in Info for chests and could reduce the classes from there.

Thanks for the review!

Rartrin commented on patches/tModLoader/Terraria/ModLoader/IO/TileIO.cs:
```diff
@@ -265,53 +458,93 @@ internal struct TileTables
-					if (pendingFrameID < 0) {
-						pendingFrameID = unloadedTiles.pendingInfos.Count;
-						unloadedTiles.pendingInfos.Add(info);
+					if (tile.type == PendingDresser) {
```
`else if` perhaps?

Solxanich replied:
Agreed. All order-sensitive stuff is dealt with, so good catch.

Solxanich commented:
Going to toss this here now, as a summary, since part 3 scope is nixed as already covered and no point in having an issue anymore:

1) Refactor across for UnloadedInfo instances and related positionMaps so that can reduce further the repeat code in UnloadedSystemsWorld and related
2) Refactor UnloadedItems to take advantage of the changes made in unloaded tile rework
3) Aim to resolve #302

Solxanich commented:
To dos for current branch based on insight gained from CB's early review comments, and subsequent revealing of opportunities:
1) fix indexOf for positionindexer per original unloadedtileinfo
2) switch position maps datatype to ordered dictionary to explicit order, rework FloorGetValue to use binary search instead of enumeration
3) realize that can calculate chest style during load instead of retrieving, thus eliminating legacy load flag and need for flags2; further can use chest style of placeholder to store chest style instead of custom data, thus trimming that down to simpler.
4) double check why pending tile variants and can't just grab variants flags when needed?
5) switch the code in pending container handling to change upcoming types to be considered handled instead of looking back
6) Create a flow chart of what TileIO+UnloadedModTilesSystem does so can figure out how to make it understandable

Solxanich commented:
Additional changes todo after discussion with CB:
1) Prevent hammering of unloadedTile to prevent frame data issues
2) refactor ModTile unloadedChest, unloaded dresser into ModTile
3) Use a calculated, overrideable field (unloadedType) in ModTile to determine variants as part of the og tile. Then can just copy-paste the field from og to placeholder. Use a quick constant byte to indicate what each means.
4) Eliminate chestInfos as its redundant after all changes made; there is nothing unique compared to standard tile
4) Swap all dicts to the appropriate combination of list, array, structure for performance; in particular, the tileTables in TileIO being a big red flag 
5) UnloadedPosIndexing should be static; its just methods, doesn't need identity
6) Eliminate the last artifact of pending flags from the old frame-based system. Store FullName to a field on UnloadedTile, and use that for the UnloadedInfo indexing check in PosIndexing, and rebuild positionMaps each load. Simplify TileTables as possible.
7) Rename fallbackType to possibly "vanillaFallbackOnModDeletion" or "vanillaReplacementType", and implement in ModTile the field to actually get it.

And anything else I see after doing all that. Move to draft.

Solxanich commented:
I've addressed the vast majority of feedback and made changes accordingly.
The following feedback I haven't addressed as written for following reasons:
1) UnloadedPosIndexing, if we were to support multiplayer run-time tile identification would imply that this should remain non-static to handle multiple different identification requests, assuming I understand static correctly
2) I was not able to fully eliminate the dictionairies in TileIO tileTables. I've brought it down from 9 dictionairies to 3. The main issue is that the tile sweep on load time creates a situation where the key's appear many to one, in addition to out of order. I believe dictionairies are appropriate here.
3) there is a few spots where refactoring could be done, but not at my current knowledge level of C#. I've flagged the three locations I've identified with comments. 
4) Added a method for getting tile placeholder type instead of a property on TileIO, based on passing string and subsequent lookup. Mod authors shouldn't need access to specify, as could lead to unpredictable behavior with current code design.

Solxanich commented:
Finished running testing; verified legacy load compatibility, and verified all methods work as expected.
RE: ready for review.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedTilesSystem.cs:
```diff
@@ -5,113 +5,49 @@
 		public override TagCompound SaveWorldData() {
 			return new TagCompound {
-				["list"] = infos.Select(info => info?.Save() ?? new TagCompound()).ToList()
+				["tileInfos"] = tileInfos.Select(info => info?.Save() ?? new TagCompound()).ToList(),
```
null infos are allowed?

direwolf420 replied:
Yes, the length of the list of infos never changes, "recovered" data is just nulled out. this keeps indexes consistent

Solxanich replied:
Thanks for responding Dire - Expounding a bit, the only usage is having fixed indices being stored in the mapping. 
The nulled entries "should" then be reusable as posIndexing should use null entries instead of adding net new.

Solxanich replied:
The entries shouldn't be null anymore, as tracking indexes are all run time now.
I've kept the null check for now, but it *should* be safe to remove, assuming everything works.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedTilesSystem.cs:
```diff
@@ -5,113 +5,49 @@
-
-				int type = ModContent.TryFind(modName, name, out ModTile tile) ? tile.Type : 0;
+			// Process tiles
+			tileInfos.AddRange(TileIO.tileInfos);
```
I think `TileIO` should have a more active interface with this class. Rather than merging two separate lists in this hook

Solxanich replied:
Class is gone, so not applicable.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/IO/TileIO.cs:
```diff
@@ -12,106 +12,261 @@ namespace Terraria.ModLoader.IO
+				}).ToList(),
+
+				["data"] = ms.ToArray(), // Array of locational-specific data
+			};
```
100/100 good stuff

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/IO/TileIO.cs:
```diff
@@ -12,106 +12,261 @@ namespace Terraria.ModLoader.IO
+
+					tabData.fallbackID = fallbackType;
+
+					UpdateUnloaded pendingInfo = new UpdateUnloaded(wallInfos, WallsContext);
```
this kind of temporary helper class is not a typical C# pattern. I'd recommend using static methods when possible.

Solxanich replied:
Also not applicable anymore.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/IO/TileIO.cs:
```diff
@@ -12,106 +12,261 @@ namespace Terraria.ModLoader.IO
+				if (tabData.id == 0) {
+					tabData.id = ModContent.Find<ModWall>("ModLoader/UnloadedWall").Type;
+					tabData.modName = modName;
+					tabData.name = name;
```
should `tabData` have an `unloadedInfo` field instead?

Solxanich replied:
Not applicable anymore.

Chicken-Bones reviewed (changes requested):
Good improvements since last time. I'll admit, I ran out of brain before I finished this review, but...

What are your thoughts on not representing unloaded data in the save file? That avoids having to merge newly unloaded tiles/walls with previously unloaded ones, and avoids having to handle a partial restoration of previously unloaded tiles/walls from a different save format.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedPosIndexing.cs:
```diff
@@ -0,0 +1,70 @@
+				if (posMap[split].posID <= this.posID) {
+					minimum = split;
+				}
+
+				if (posMap[split + 1].posID > this.posID) {
+					maximum = split;
+				}
+
```
Given we are looking for the interval from posMap[split] to posMap[split + 1] containing posID, it may be more effective to separate it out as it decreases the amount of iterations for elements closer to center. IIRC it should be net beneficial, as >80% of domain benefits.

An example being that if the posID is in the dead middle (say 50.1 / 100) - if I use if else, it would find the minimum is 50, but then take another few checks to determine the maximum is also 50.

Solxanich commented:
@Chicken-Bones - per our discussion, I've gone a whole new route and rewrote TileIO with just doing tiles for now. Can you take a look, before I start thinking about generics for doing walls using the same methods?

@direwolf420  - the PosIndexer has been generalized and should be closer to what you were looking for - let me know if other methods to add.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedTile.cs:
```diff
@@ -1,37 +1,138 @@
+			var infos = modSystem.tileInfos;
+
+			if (infoID < infos.Count) {
+				var info = infos[infoID];
```
I've simplified this to be a fair few less lines. 
PosIndexer.GetKeyFromPos() returns the index for a given position, in the provided position map, which is more general and usable elsewise, but still simplified.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedPosIndexing.cs:
```diff
@@ -0,0 +1,70 @@
+		public void MapPosToInfo(List<UnloadedInfo> infos, List<TileIO.posMap> posMap, List<TileIO.posMap> prevPosMap = null, UnloadedInfo info = null)  {
+			ushort pendingID = 0;
+			if (info == null && prevPosMap.Count > 0) {
+				pendingID = (ushort) (FloorGetValue(prevPosMap.ToArray()) + infos.Count);
```
Fixed. Misunderstood the request while discussing. 
Uses a list while generating, and converts list to an array for run time.

Performance is slightly slower on saving, didn't notice any visual indication of performance drop during run time when using 
[ReplaceTilesCommand.zip](https://github.com/tModLoader/tModLoader/files/6031450/ReplaceTilesCommand.zip)

Should be good now.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedTile.cs:
```diff
@@ -1,37 +1,138 @@
+		internal bool isSolid;
+		internal bool isSemi;
+		internal bool isChest;
+		internal bool isDresser;
```
TODO self: probably subclass, can unique-ify the textures so they can be distinguished, such as chests and dressers

Solxanich commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedInfo.cs:
```diff
@@ -0,0 +1,39 @@
+		}
+
+		public override int GetHashCode() {
+			int hash = name.GetHashCode() + modName.GetHashCode();
```
The concern here is collision, correct?
Prime multiply and add only holds for uniqueness on counting numbers iirc. Given GetHashCode can return negative, that won't be sufficient to eliminate collision. I can abs, but that still isn't perfect.  
XOR has same negative issue. It prevents overflow so I tend to prefer it, so I'll implement with absolute as that's better than adding, but I'm open to better ideas.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/PosData.cs:
```diff
@@ -0,0 +1,150 @@
+			while (maximum - minimum > 1) {
+				int split = (minimum + maximum) / 2;
+
+				if (posMap[split].pos <= pos)
```
Reverse this, again per original review, its wrong for an interval search as the exit condition takes wasted cycles to reach

Solxanich commented on patches/tModLoader/Terraria/ModLoader/IO/ModEntry.cs:
```diff
@@ -1,51 +1,41 @@
+			modName = block.Mod.Name;
+			name = block.Name;
+			vanillaReplacementType = block.vanillaFallbackOnModDeletion;
+			throw new NotImplementedException();// unloadedType = ?
```
Unloaded type was implemented, there was a method in TileIO basic to calc it.
Why is this here?

Solxanich commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedTile.cs:
```diff
@@ -2,40 +2,25 @@
-					player.cursorItemIconText = $"{info.modName}: {info.name}";
-				}
+			var info = TileIO.tileEntries[TileIO.unloadedTileLookup.Lookup(i, j)];
+			if (info != null) {
```
Need to test this

Solxanich commented on patches/tModLoader/Terraria/ModLoader/IO/TileIO_Basic.cs:
```diff
@@ -7,206 +7,81 @@
 namespace Terraria.ModLoader.IO
 {
 	internal static partial class TileIO {
-		static string tileEntries = "tileMap";
```
Why is this deleted with no replacement?

Solxanich reviewed (commented):
Some to dos.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/IO/WallEntry.cs:
```diff
@@ -0,0 +1,16 @@
+	{
+		public static Func<TagCompound, WallEntry> DESERIALIZER = tag => DeserializeTag(tag);
+
+		public static WallEntry DeserializeTag(TagCompound tag) {
```
This *probably* could be generalized and moved up to ModEntry, but I missed that when doing it.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/PosData.cs:
```diff
@@ -0,0 +1,166 @@
+		public static int FindIndex<T>(this PosData<T>[] posMap, int pos) {
+			if (posMap.Length == 0) {
+				throw new ArgumentException($"Can't find the index in an empty posMap. Please verify map is non-empty before calling.");
+			}
```
please remove this condition. I updated the iteration to return -1 when `pos` is < the first entry in the map (or the map is empty)

This works naturally with the builder, as there's essentially a phantom element at `-1` with the value of `default(T)` this means you don't ever have to check that the map is non-empty.

Conceptually, an empty map means that the entire world is filled with `default(T)`, which is a sensible thing to work with

bvcbvc commented:
i understand none of this, but yes unloaded tiles deserve this treatment

Solxanich commented on patches/tModLoader/Terraria/ModLoader/PosData.cs:
```diff
@@ -116,26 +130,44 @@ public static class PosData
+		/// General purpose lookup function. Always returns a value (even if that value is `default`).
+		/// See <see cref="PosData{T}.OrderedSparseLookupBuilder.OrderedSparseLookupBuilder(int, bool, bool)"/>for more info
+		/// </summary>
+		public static T Lookup<T>(this PosData<T>[] posMap, int x, int y) => posMap.Lookup(CoordsToPos(x, y));
```
Modders can always directly call posData.Find() if using default values as valid values and use that to check if data is valid.

Solxanich reviewed (commented):
Review commit 1.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedWall.cs:
```diff
@@ -7,7 +7,7 @@ public class UnloadedWall : ModWall
 
 		public override void SetDefaults() {
-			TileIO.IsUnloadedWall.Add(Type);
+			TileIO.Tiles.unloadedTypes.Add(Type);
```
Double check

Solxanich replied:
Double check if should be Walls

Solxanich commented on patches/tModLoader/Terraria/ModLoader/IO/ModEntry.cs:
```diff
@@ -9,26 +9,28 @@ public abstract class ModEntry : TagSerializable
-		public abstract string DefaultUnloadedType { get; }
-
-		internal virtual void SetData<B>(B block) where B : ModBlockType {
+		protected ModEntry(ModBlockType block) {
```
Double check how got constructors to work in generics?

Solxanich replied:
Didn't just used inheritance.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/IO/TileIO_Basic.cs:
```diff
@@ -6,265 +7,243 @@
-		// todo, resize arrays or content tags
-		public static List<ushort> IsUnloadedTile = new List<ushort>();
-		public static List<ushort> IsUnloadedWall = new List<ushort>();
+		public abstract class IOImpl<TBlock, TEntry> where TBlock : ModBlockType where TEntry : ModEntry
```
Privacy mismatch. Internal

Solxanich commented on patches/tModLoader/Terraria/ModLoader/IO/TileIO_Basic.cs:
```diff
@@ -1,3 +1,4 @@
+using System;
```
Figure out where call:
TileIO.LoadNewWorld

Solxanich replied:
Was unneeded. LoadEntries assigns CreateEntries off the get go, and unloadedEntryLookup is assigned to by builder.Build in LoadData before it is used.

Solxanich reviewed (commented):
Review commit 2; will need to look at TileIO basic in VS as its non-trivial to look through

Solxanich commented on patches/tModLoader/Terraria/ModLoader/PosData.cs:
```diff
@@ -23,16 +38,13 @@ public class OrderedSparseLookupBuilder
-			public void ClusteredAdd(int pos, T value) {
-				if (pos <= last.pos)
-					throw new ArgumentException($"Must build in ascending index order. Prev: {last.pos}, pos: {pos}");
+				if (compressEqualValues && !insertDefaultEntries && pos > last.pos+1) {
```
Yeah, no ! on insertDefaultEntries.
